File: 0120-triangle/0120-triangle.java
class Solution {
    
//     public int memorization(int i,int j,int n,int [][]mem,List<List<Integer>>triangle){
//         if(i==n)return 0;

//         if(mem[i][j]!=-1)return mem[i][j];
        
//         int left=triangle.get(i).get(j)+memorization(i+1,j,n,mem,triangle);
//         int right=triangle.get(i).get(j)+memorization(i+1,j+1,n,mem,triangle);
//         return mem[i][j]=Math.min(left,right);
        
//     }
    
    public int minimumTotal(List<List<Integer>> triangle) {
        int n=triangle.size();
        // int [][]dp=new int[n][n];
        // for(int[]d:dp)
            // Arrays.fill(d,-1);
        
        // return memorization(0,0,n,dp,triangle);

        // for(int j=0;j<n;j++)dp[n-1][j]=triangle.get(n-1).get(j);
        
        // for(int i=n-2;i>=0;i--){
        //     for(int j=0;j<=i;j++){
        //         dp[i][j]=Math.min(dp[i+1][j]+triangle.get(i).get(j),dp[i+1][j+1]+triangle.get(i).get(j));
        //     }
        // }
        // return dp[0][0];
        
        int []nextRow=new int[n+1];

        for(int i=n-1;i>=0;i--){
            for(int j=0;j<triangle.get(i).size();j++){
                
                int left=nextRow[j]+triangle.get(i).get(j);
                int right=nextRow[j+1]+triangle.get(i).get(j);
                nextRow[j]=Math.min(left,right);
                
            }
        }
        
        return nextRow[0];
        
    }
}

File: 0120-triangle/0120-triangle.cpp
class Solution {
public:
    int minimumTotal(vector<vector<int>>& triangle) {
        
        // vector<vector<int>>dp(triangle.size(),vector<int>(triangle.size()));
        // dp[0][0]=triangle[0][0];
        // for(int i=1;i<triangle.size();i++){
        //     for(int j=0;j<triangle[i].size();j++){
        //         if(j==0)dp[i][j]=dp[i-1][j]+triangle[i][j];
        //         else if(j==triangle[i].size()-1)dp[i][j]=dp[i-1][j-1]+triangle[i][j];
        //         else dp[i][j]=triangle[i][j]+min(dp[i-1][j-1],dp[i-1][j]);
        //     }
        // }
        // return *min_element(dp[triangle.size()-1].begin(),dp[triangle.size()-1].end());
        int n=triangle.size();
        vector<int>currRow(n+1,0);
        for(int i=n-1;i>=0;i--){
            for(int j=0;j<triangle[i].size();j++){
                currRow[j]=min(currRow[j],currRow[j+1])+triangle[i][j];
            }
        }
        return currRow[0];
        
    }
};